### PR TITLE
plugin/rewrite: fix domain length validation

### PR DESCRIPTION
**Pull request by rdrozhdzh**
### 1. Why is this pull request needed and what does it do?
fix domain length validation in rewrite plugin
### 2. Which issues (if any) are related?
#2992
### 3. Which documentation changes (if any) need to be made?
none
### 4. Does this introduce a backward incompatible change or deprecation?
no

### CODE REVIEW

**Review by miekg — COMMENTED** [review]
lgtm, but do we really need a test for this? We can hope dns.IsDomainName is tested well enough

**Comment by stp-ip** [comment]
/lgtm
But is hope a good strategy? :smiling_imp:

**Comment by rdrozhdzh** [comment]
Yes, I'm sure `dns.IsDomainName()` is tested well. I'd rather want to make sure that rewrite plugin doesn't do anything freaky in addition to checking with `dns.IsDomainName()` like it was before my fix.

Though, I can remove my unit test if you wish. I just added it for confirmation that my fix works as expected.

**Comment by miekg** [comment]
I think it's a rather convoluted way to test if IsDomainName does what it needs to do.

More generally it would be good to test this, but then for all plugins which is is way harder to setup.

File: plugin/rewrite/rewrite.go [hunk 1]
```diff
@@ -44,7 +44,8 @@ func (rw Rewrite) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.Msg
 	for _, rule := range rw.Rules {
 		switch result := rule.Rewrite(ctx, state); result {
 		case RewriteDone:
-			if !validName(state.Req.Question[0].Name) {
+			_, validName := dns.IsDomainName(state.Req.Question[0].Name)
```
Comment (miekg) [inline plugin/rewrite/rewrite.go]: the `if` can be pulled up, and just using `ok` is more go like:
~~~
if _, ok := dns.IsDomainName; !ok {

}
~~~
also the `x :=` seemed superfluous and why are we logging this *and* return the error, seem we can drop the `log.Errrof` here as well

Comment (rdrozhdzh) [reply plugin/rewrite/rewrite.go]: Regarding `x :=` - we we need to save the name before overwriting `Question[0]` with `originalQuestion`, the name is needed to construct the error message to be returned by `ServeDNS()`. I'm going to update the code

**Comment by rdrozhdzh** [comment]
Seems the build failure is not related to my change. Not sure if I can re-run the tests
```
--- FAIL: TestRequestScrubExtra (0.00s)
    request_test.go:102: Want scrub to not set truncated bit
=== RUN   TestRequestScrubExtraEdns0
--- FAIL: TestRequestScrubExtraEdns0 (0.00s)
    request_test.go:124: Want scrub to not set truncated bit
=== RUN   TestRequestScrubExtraRegression
--- FAIL: TestRequestScrubExtraRegression (0.00s)
    request_test.go:150: Want scrub to not set truncated bit
=== RUN   TestTruncation
--- PASS: TestTruncation (0.08s)
=== RUN   TestRequestScrubAnswerExact
--- PASS: TestRequestScrubAnswerExact (0.00s)
=== RUN   TestRequestMatch
--- PASS: TestRequestMatch (0.00s)
=== RUN   TestRequestClear
--- PASS: TestRequestClear (0.00s)
FAIL
coverage: 40.1% of statements
FAIL	github.com/coredns/coredns/request	0.092s
```

**Comment by chrisohaver** [comment]
re-running tests...

**Comment by miekg** [comment]
It is unrelated. Bit strange this would fail now, but will fix.

**Comment by miekg** [comment]
please remove the test, other then that lgtm

**Pull request merged**